setup/setup.js

```diff
@@ -0,0 +1,108 @@
+"use strict";
+
+const fs = require("fs");
+const path = require("path");
+const root = process.cwd();
+const node_modulesFolder = path.resolve(root, "node_modules");
+const webpackDependencyFolder = path.resolve(root, "node_modules/webpack");
+
+function setup() {
+	return checkSymlinkExistsAsync()
+		.then(hasSymlink => {
+			if (!hasSymlink) {
+				return ensureYarnInstalledAsync().then(() => {
+					return runSetupAsync().then(() => {
+						return checkSymlinkExistsAsync();
+					});
+				});
+			}
+		})
+		.then(message => {
+			process.exitCode = 0;
+		})
+		.catch(e => {
+			console.error(e);
+			process.exitCode = 1;
+		});
+}
+
+function runSetupAsync() {
+	return exec("yarn", ["install"], "Install dependencies")
+		.then(() => exec("yarn", ["link"], "Create webpack symlink"))
+		.then(() => exec("yarn", ["link", "webpack"], "Link webpack into itself"));
+}
+
+function checkSymlinkExistsAsync() {
+	return new Promise((resolve, reject) => {
+		if (
+			fs.existsSync(node_modulesFolder) &&
+			fs.existsSync(webpackDependencyFolder) &&
+			fs.lstatSync(webpackDependencyFolder).isSymbolicLink()
+		) {
+			resolve(true);
+		} else {
+			resolve(false);
+		}
+	});
+}
+
+function ensureYarnInstalledAsync() {
+	var semverPattern = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(\.(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*)?(\+[0-9a-zA-Z-]+(\.[0-9a-zA-Z-]+)*)?$/;
+	return execGetOutput("yarn", ["-v"], "Check yarn version")
+		.then(stdout => semverPattern.test(stdout), () => false)
+		.then(hasYarn => hasYarn || installYarnAsync());
+}
+
+function installYarnAsync() {
+	return exec("npm", ["install", "-g", "yarn"], "Install yarn");
+}
+
+function exec(command, args, description) {
+	console.log(`Setup: ${description}`);
+	return new Promise((resolve, reject) => {
+		let cp = require("child_process").spawn(command, args, {
```

Comment:
I would recommend using `execa`. It is already promised, it is more stable on `windows/freebsd`, it is have simple api and increased child_proccess buffer and etc

Reply:
Good idea, but sadly we can't use any dependency, since the setup script runs before dependencies are installed.

Reply:
Exactly, I was just looking for `execa` in the Node API, but it's a separate package.